website/project/model.py

```diff
@@ -3729,3 +3810,125 @@ def _on_reject(self, user, token):
+
+
+class DraftRegistration(AddonModelMixin, StoredObject):
+
+    is_draft_registration = True
+
+    _id = fields.StringField(primary=True, default=lambda: str(ObjectId()))
+
+    datetime_initiated = fields.DateTimeField(auto_now_add=True)
+    datetime_updated = fields.DateTimeField(auto_now=True)
+    # Original Node a draft registration is associated with
+    branched_from = fields.ForeignField('node')
+
+    initiator = fields.ForeignField('user')
+
+    # Dictionary field mapping question id to a question's comments and answer
+    # {
+    #   <qid>: {
+    #     'comments': [{
+    #       'user': {
+    #         'id': <uid>,
+    #         'name': <name>
+    #       },
+    #       value: <value>,
+    #       lastModified: <datetime>
+    #     }],
+    #     'value': <value>
+    #   }
+    # }
+    registration_metadata = fields.DictionaryField(default=dict)
+    registration_schema = fields.ForeignField('metaschema')
+    registered_node = fields.ForeignField('node')
+
+    approval = fields.ForeignField('draftregistrationapproval', default=None)
+
+    # Dictionary field mapping extra fields defined in the MetaSchema.schema to their
+    # values. Defaults should be provided in the schema (e.g. 'paymentSent': false),
+    # and these values are added to the DraftRegistration
+    flags = fields.DictionaryField()
+
+    notes = fields.StringField()
+
+    def __init__(self, *args, **kwargs):
+        super(DraftRegistration, self).__init__(*args, **kwargs)
+        meta_schema = self.registration_schema or kwargs.get('registration_schema')
+        if meta_schema:
+            schema = meta_schema.schema
+            if not self.registration_schema:
+                flags = schema.get('flags', {})
+                for flag, value in flags.iteritems():
+                    self.flags[flag] = value
+
+    @property
+    def requires_approval(self):
+        return self.registration_schema.requires_approval
+
+    @property
+    def is_pending_review(self):
+        return self.approval.is_pending_approval if (self.requires_approval and self.approval) else False
+
+    @property
+    def is_approved(self):
+        if self.requires_approval and not self.approval:
+            return False
+        elif self.requires_approval and self.approval:
+            return self.approval.is_approved
+        else:
+            return True
+
+    def update_metadata(self, metadata):
+        changes = []
+        for question_id, value in metadata.iteritems():
+
+            old_value = self.registration_metadata.get(question_id)
+
+            if not old_value or old_value.get('value') != value.get('value'):
+                changes.append(question_id)
+
+        self.registration_metadata.update(metadata)
+
+        project_signals.draft_edited.send(changes)
+        self.after_edit(changes)
+        return changes
+
+    def after_edit(self, changes):
+
+        if changes:
+            self.flags.update({
```

Comment:
As I understand it, self.flags is an arbitrary dictionary to support schema specific dictionaries... but the parent `DraftRegistration` method sets a number of flags that are very specific to one kind of registration.

At minimum we should document the available flags and there intended behavior- since strictly speaking, the parent class is building in behavior intended only for one specific use case.

Discussion question: Are there any scenarios where the presence of these flags (on a non-prereg schema) could lead to unintended behavior? If so, has this been incorporated into the QA testing plan?